File: svix-cli/src/cli_types/message_attempt.rs
use chrono::{DateTime, Utc};
use clap::Args;
use svix::{
    api,
    api::{MessageStatus, StatusCodeClass},
};

#[derive(Args, Clone)]
pub struct MessageAttemptListByEndpointOptions {
    /// Limit the number of returned items
    #[arg(long)]
    pub limit: Option<i32>,
    /// The iterator returned from a prior invocation
    #[arg(long)]
    pub iterator: Option<String>,
    /// Filter response based on the status of the attempt: Success (0), Pending (1), Failed (2), or Sending (3)
    #[arg(long)]
    pub status: Option<MessageStatus>,
    /// Filter response based on the HTTP status code
    #[arg(long)]
    pub status_code_class: Option<StatusCodeClass>,
    /// Filter response based on the channel
    #[arg(long)]
    pub channel: Option<String>,
    /// Filter response based on the tag
    #[arg(long)]
    pub tag: Option<String>,
    /// Only include items created before a certain date
    #[arg(long)]
    pub before: Option<DateTime<Utc>>,
    /// Only include items created after a certain date
    #[arg(long)]
    pub after: Option<DateTime<Utc>>,
    /// When `true` attempt content is included in the response
    #[arg(long)]
    pub with_content: Option<bool>,
    /// When `true`, the message information is included in the response
    #[arg(long)]
    pub with_msg: Option<bool>,
    /// Filter response based on the event type
    #[arg(long)]
    pub event_types: Option<Vec<String>>,
}

impl From<MessageAttemptListByEndpointOptions> for api::MessageAttemptListByEndpointOptions {
    fn from(
        MessageAttemptListByEndpointOptions {
            limit,
            iterator,
            status,
            status_code_class,
            channel,
            tag,
            before,
            after,
            with_content,
            with_msg,
            event_types,
        }: MessageAttemptListByEndpointOptions,
    ) -> Self {
        Self {
            limit,
            iterator,
            status,
            status_code_class,
            channel,
            tag,
            before: before.map(|dt| dt.to_rfc3339()),
            after: after.map(|dt| dt.to_rfc3339()),
            with_content,
            with_msg,
            event_types,
        }
    }
}

#[derive(Args, Clone)]
pub struct MessageAttemptCountByEndpointOptions {
    /// Filter response based on the status of the attempt: Success (0), Pending (1), Failed (2), or Sending (3)
    #[arg(long)]
    pub status: Option<MessageStatus>,
    /// Filter response based on the HTTP status code
    #[arg(long)]
    pub status_code_class: Option<StatusCodeClass>,
    /// Filter response based on the channel
    #[arg(long)]
    pub channel: Option<String>,
    /// Filter response based on the tag
    #[arg(long)]
    pub tag: Option<String>,
    /// Only include items created before a certain date
    #[arg(long)]
    pub before: Option<DateTime<Utc>>,
    /// Only include items created after a certain date
    #[arg(long)]
    pub after: Option<DateTime<Utc>>,
    /// Filter response based on the event type
    #[arg(long)]
    pub event_types: Option<Vec<String>>,
}

#[derive(Args, Clone)]
pub struct MessageAttemptListByMsgOptions {
    /// Limit the number of returned items
    #[arg(long)]
    pub limit: Option<i32>,
    /// The iterator returned from a prior invocation
    #[arg(long)]
    pub iterator: Option<String>,
    /// Filter response based on the status of the attempt: Success (0), Pending (1), Failed (2), or Sending (3)
    #[arg(long)]
    pub status: Option<MessageStatus>,
    /// Filter response based on the HTTP status code
    #[arg(long)]
    pub status_code_class: Option<StatusCodeClass>,
    /// Filter response based on the channel
    #[arg(long)]
    pub channel: Option<String>,
    /// Filter response based on the tag
    #[arg(long)]
    pub tag: Option<String>,
    /// Filter the attempts based on the attempted endpoint
    #[arg(long)]
    pub endpoint_id: Option<String>,
    /// Only include items created before a certain date
    #[arg(long)]
    pub before: Option<DateTime<Utc>>,
    /// Only include items created after a certain date
    #[arg(long)]
    pub after: Option<DateTime<Utc>>,
    /// When `true` attempt content is included in the response
    #[arg(long)]
    pub with_content: Option<bool>,
    /// Filter response based on the event type
    #[arg(long)]
    pub event_types: Option<Vec<String>>,
}

impl From<MessageAttemptListByMsgOptions> for api::MessageAttemptListByMsgOptions {
    fn from(
        MessageAttemptListByMsgOptions {
            limit,
            iterator,
            status,
            status_code_class,
            channel,
            tag,
            endpoint_id,
            before,
            after,
            with_content,
            event_types,
        }: MessageAttemptListByMsgOptions,
    ) -> Self {
        Self {
            limit,
            iterator,
            status,
            status_code_class,
            channel,
            tag,
            endpoint_id,
            before: before.map(|dt| dt.to_rfc3339()),
            after: after.map(|dt| dt.to_rfc3339()),
            with_content,
            event_types,
        }
    }
}

#[derive(Args, Clone)]
pub struct MessageAttemptListAttemptedMessagesOptions {
    /// Limit the number of returned items
    #[arg(long)]
    pub limit: Option<i32>,
    /// The iterator returned from a prior invocation
    #[arg(long)]
    pub iterator: Option<String>,
    /// Filter response based on the channel
    #[arg(long)]
    pub channel: Option<String>,
    /// Filter response based on the message tags
    #[arg(long)]
    pub tag: Option<String>,
    /// Filter response based on the status of the attempt: Success (0), Pending (1), Failed (2), or Sending (3)
    #[arg(long)]
    pub status: Option<MessageStatus>,
    /// Only include items created before a certain date
    #[arg(long)]
    pub before: Option<DateTime<Utc>>,
    /// Only include items created after a certain date
    #[arg(long)]
    pub after: Option<DateTime<Utc>>,
    /// When `true` message payloads are included in the response
    #[arg(long)]
    pub with_content: Option<bool>,
    /// Filter response based on the event type
    #[arg(long)]
    pub event_types: Option<Vec<String>>,
}

impl From<MessageAttemptListAttemptedMessagesOptions>
    for api::MessageAttemptListAttemptedMessagesOptions
{
    fn from(
        MessageAttemptListAttemptedMessagesOptions {
            limit,
            iterator,
            channel,
            tag,
            status,
            before,
            after,
            with_content,
            event_types,
        }: MessageAttemptListAttemptedMessagesOptions,
    ) -> Self {
        Self {
            limit,
            iterator,
            channel,
            tag,
            status,
            before: before.map(|dt| dt.to_rfc3339()),
            after: after.map(|dt| dt.to_rfc3339()),
            with_content,
            event_types,
        }
    }
}

#[derive(Args, Clone)]
pub struct MessageAttemptListAttemptedDestinationsOptions {
    /// Limit the number of returned items
    #[arg(long)]
    pub limit: Option<i32>,
    /// The iterator returned from a prior invocation
    #[arg(long)]
    pub iterator: Option<String>,
}

impl From<MessageAttemptListAttemptedDestinationsOptions>
    for api::MessageAttemptListAttemptedDestinationsOptions
{
    fn from(
        MessageAttemptListAttemptedDestinationsOptions {
                    limit,iterator,
                }: MessageAttemptListAttemptedDestinationsOptions,
    ) -> Self {
        Self { limit, iterator }
    }
}

File: javascript/src/api_internal/messageAttempt.ts
// this file is @generated
import { CountOut, CountOutSerializer } from "../models/countOut";
import {
  MessageAttemptHeadersOut,
  MessageAttemptHeadersOutSerializer,
} from "../models/messageAttemptHeadersOut";
import { MessageStatus } from "../models/messageStatus";
import { StatusCodeClass } from "../models/statusCodeClass";
import { HttpMethod, SvixRequest, SvixRequestContext } from "../request";

export interface MessageAttemptCountByEndpointOptions {
  /** Filter response based on the status of the attempt: Success (0), Pending (1), Failed (2), or Sending (3) */
  status?: MessageStatus;
  /** Filter response based on the HTTP status code */
  statusCodeClass?: StatusCodeClass;
  /** Filter response based on the channel */
  channel?: string;
  /** Filter response based on the tag */
  tag?: string;
  /** Only include items created before a certain date */
  before?: Date | null;
  /** Only include items created after a certain date */
  after?: Date | null;
  /** Filter response based on the event type */
  eventTypes?: string[];
}

export class MessageAttempt {
  public constructor(private readonly requestCtx: SvixRequestContext) {}

  /** Like `v1.message-attempt.list-by-endpoint` but returning a count only. */
  public countByEndpoint(
    appId: string,
    endpointId: string,
    options?: MessageAttemptCountByEndpointOptions
  ): Promise<CountOut> {
    const request = new SvixRequest(
      HttpMethod.GET,
      "/api/v1/app/{app_id}/attempt/endpoint/{endpoint_id}/count"
    );

    request.setPathParam("app_id", appId);
    request.setPathParam("endpoint_id", endpointId);
    request.setQueryParam("status", options?.status);
    request.setQueryParam("status_code_class", options?.statusCodeClass);
    request.setQueryParam("channel", options?.channel);
    request.setQueryParam("tag", options?.tag);
    request.setQueryParam("before", options?.before);
    request.setQueryParam("after", options?.after);
    request.setQueryParam("event_types", options?.eventTypes);

    return request.send(this.requestCtx, CountOutSerializer._fromJsonObject);
  }

  /** Calculate and return headers used on a given message attempt */
  public getHeaders(
    appId: string,
    msgId: string,
    attemptId: string
  ): Promise<MessageAttemptHeadersOut> {
    const request = new SvixRequest(
      HttpMethod.GET,
      "/api/v1/app/{app_id}/msg/{msg_id}/attempt/{attempt_id}/headers"
    );

    request.setPathParam("app_id", appId);
    request.setPathParam("msg_id", msgId);
    request.setPathParam("attempt_id", attemptId);

    return request.send(
      this.requestCtx,
      MessageAttemptHeadersOutSerializer._fromJsonObject
    );
  }
}
